### PR TITLE
librbd: do not instantiate templates while building tests

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by tchaikov** [comment]
@dillaman i wish i knew the right way the common practice of librados tests as you guys. but it's not obvious to me how we can resolve the issue of duplicated classes definition such as `librbd::AsyncObjectThrottle<librbd::ImageCtx>` using the `MockImageCtx` in anonymous namespace? each inclusion of `librbd/AsyncObjectThrottle.cc` instantiate a new `librbd::AsyncObjectThrottle<librbd::ImageCtx>` instantiation in the including compilation unit.

**Comment by dillaman** [comment]
@tchaikov Ah -- I misinterpreted the issue you were trying to fix. It sounds like the armhf GCC is not tagging the explicitly instantiated template symbols as weak like the other GCC compilers. I think I'd rather just ```#ifdef SOMtiation such that they are only activated when building librbd -- and not when pulled in via a test.

**Comment by tchaikov** [comment]
@dillaman updated and repushed.

**Comment by tchaikov** [comment]
test failure is addressed by #14929

**Comment by dillaman** [comment]
@tchaikov Has this been tested? Just wondering why only a few classes need this workaround instead of all the classes that have mock tests.

**Comment by tchaikov** [comment]
@dillaman tested by compiling unittest_librbd using gcc-armhf.

**Comment by dillaman** [comment]
@tchaikov Do you have any idea why the DisableFeaturesRequest needed the #ifndef? Is there a way for me to run a cross-compile on my local machine to look into this?

**Comment by tchaikov** [comment]
@dillaman because "src/test/librbd/operation/test_mock_DisableFeaturesRequest.cc" includes `librbd/operation/DisableFeaturesRequest.cc`. yes. you can create a toolchain.cmake file like https://github.com/ceph/ceph/pull/14881#issuecomment-298301640, and launch cmake like
```sh
cmake -DCMAKE_TOOLCHAIN_FILE=~/toolchain-arm-linux-gnueabi.cmake
```
after installing all dev packages, on debian (or its derivatives), one can install them like
```sh
dpkg --add-architecture armhf
apt-get update
apt-get install crossbuild-essential-armhf
apt-get install libudev-dev:armhf
...
```

**Comment by dillaman** [comment]
@tchaikov Yes -- understand that the mock test includes the implementation, but so do all the other mock tests so I was wondering what was so special about just that one test:

> $ git grep -E "#include.*\.cc" | wc -l
> 48

**Comment by tchaikov** [comment]
@dillaman yeah, not sure why =(. i was just fixing anything linker was complaining about.

**Comment by dillaman** [comment]
@tchaikov Ack -- thx, I'll try to look into it some more.

**Comment by tchaikov** [comment]
thanks, i am assigning this PR to you for future investigation.

**Comment by dillaman** [comment]
@tchaikov Did you have to update your apt sources.list somehow?  Neither "apt-get install libudev-dev:armhf" nor "mk-build-deps ..." works

**Comment by tchaikov** [comment]
@dillaman  did you run "apt-get update" after "dpkg --add-architecture armhf"?

**Comment by dillaman** [comment]
@tchaikov Yup -- status shows it fails to download the package lists for armhf. That's why I was wondering if you had to do something w/ your source list.

**Comment by dillaman** [comment]
@tchaikov Are you cross compiling from an arm64 box? It looks like all the arm packages are in a totally different repo @ http://ports.ubuntu.com/ubuntu-ports/dists/xenial/main/

**Comment by tchaikov** [comment]
@dillaman i am cross-compiling from an amd64 box. btw, i am using debian sid. will post you my sources.list tomorrow. maybe ubuntu/xenial is different.

**Comment by dillaman** [comment]
@tchaikov I'll try debian -- thanks

**Comment by tchaikov** [comment]
@dillaman thank you! i will also try to `nm` other other mock tests to understand why my fix works.

**Comment by dillaman** [comment]
@tchaikov I still wasn't able to get cross-compiling working under Debian unstable due to all the package dependencies. I am currently just running a build under a qemu-arm-static chroot environment.

**Comment by tchaikov** [comment]
@dillaman i updated http://tracker.ceph.com/issues/18938#note-46 to answer your question at https://github.com/ceph/ceph/pull/14891#issuecomment-298926973, 

tl;dr: `AsyncRequest` defines key function. so does its derived classes. and gcc-armhf does not put type_info into COMDAT sections to comply the ARM's C++ ABI spec.

File: src/librbd/operation/DisableFeaturesRequest.cc [hunk 1]
```diff
@@ -636,4 +636,6 @@ Context *DisableFeaturesRequest<I>::handle_finish(int r) {
 } // namespace operation
 } // namespace librbd
 
+#ifndef TEST_F
```
Comment (dillaman) [inline src/librbd/operation/DisableFeaturesRequest.cc]: This doesn't appear to be necessary -- I successfully built and linked unittest_librbd w/o this change.

Comment (tchaikov) [inline src/librbd/operation/DisableFeaturesRequest.cc]: no, it does not. updated and pushed.
